### PR TITLE
feat(curriculum): add voting system lab

### LINKED ISSUE #655: Add Maps and Sets lab to main repo
Once the lab is approved, then we can add it to main with tests

### PR DESCRIPTION
**Pull request by zairahira**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes https://github.com/freeCodeCamp/CurriculumExpansion/issues/655

### CODE REVIEW

File: curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+   - If the voter has already voted, the function should display the message `Voter <voterId> has already voted for "<option>".`
+
+   - If the voter has not voted, `voterId` should be added to the `Set` of `voters` for this option. The function should return the message  `Voter <voterId> voted for "<option>".`
+
+6. You should have a function `displayResults` that returns the poll results in the following format:
+
+```js
+Poll Results:
+Option A: N votes
```
[the remark this answers is not in the document]
Comment (zairahira) [reply curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md:44]: Is this better:

````md
```js
Poll Results:
OptionA: N votes
OptionB: N votes
.
.

/* 
sample output

Poll Results:
Turkey: 2 votes
Morocco: 1 votes
*/
```

````

[the remark this answers is not in the document]
Comment (ilenia-magoni) [reply curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md:44]: that would be more clear for sure

File: curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+   - If the `option` already exists, it should return the message `Option "<option>" already exists.`.
+
+   - If you try to add an empty option, the function should return the message `Option cannot be empty.`.
+
+4. You should have a function `vote` that accepts two parameters, `option`(the option to vote for) and `voterId` (a unique ID for the voter).
+
+5. In the `vote` function:
+
+   - If the `option` does not exist in the poll, the function should return the message `Option "<option>" does not exist.`;
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md:32]:
```suggestion
   - If the `option` does not exist in the poll, the function should return the message `Option "<option>" does not exist.`.
```

**Review by moT01 — CHANGES_REQUESTED** [review]
This is a lot of fun @zairahira 🎉 I left comments on a few minor typos below. One other issue I has was that the tests were interfering with me testing and debugging my code:

<details><summary>image</summary>

Screenshot 2024-11-22 at 10 06 17 AM

</details>

They are overwriting the things I add. I think we need to try and fix this. Can we do some reassignment to not overwrite the users polls/votes?

One other thing you could add is some user stories and tests like I mentioned on the bank account lab to have users add to their poll and make votes. e.g.

"You should have at least three options in your poll."
"Your poll should have at least three votes."

Not necessary to add that, but I think it makes it better.

**Comment by zairahira** [comment]
@moT01,

I have modified the tests to not use `poll.clear()` every time. Maybe this was interfering/overriding your solutions.

I have also added the following tests/user stories:

- You should have at least three options in your `poll`.
- Your `poll` should have at least three votes.

Let me know if this is good to go.

**Review by Dario-DC — COMMENTED** [review]
I still see the outcome of the last test in the console. Ideally, you should save a copy of the map, make your changes to test things, and then restore the map to how it was before the test.

Something like this:

```js
const pollCopy = new Map(poll);
try {
  poll.clear();
  // modify poll as needed
  // assert things
} finally {
  // the finally block is required for when the assert fails
  // otherwise you won't restore the map
  poll.clear();
  pollCopy.forEach((val, key) => {
    poll.set(key, val);
  });
}
```

File: client/i18n/locales/english/intro.json [hunk 3]
```diff
@@ -3096,7 +3096,12 @@
         ]
       },
       "cvsw": { "title": "226", "intro": [] },
-      "cvs1": { "title": "227", "intro": [] },
+      "lab-voting-system": {
+        "title": "Build a Voting System Using Maps and Sets",
```
Comment (Dario-DC) [inline client/i18n/locales/english/intro.json:3100]:
```suggestion
        "title": "Build a Voting System",
```

File: client/src/pages/learn/full-stack-developer/lab-voting-system/index.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+---
+title: Introduction to the Build a Voting System Using Maps and Sets
```
Comment (Dario-DC) [inline client/src/pages/learn/full-stack-developer/lab-voting-system/index.md:2]:
```suggestion
title: Introduction to the Build a Voting System
```

File: client/src/pages/learn/full-stack-developer/lab-voting-system/index.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+block: lab-voting-system
+superBlock: full-stack-developer
+---
+
+## Introduction to the Build a Voting System Using Maps and Sets
```
Comment (Dario-DC) [inline client/src/pages/learn/full-stack-developer/lab-voting-system/index.md:7]:
```suggestion
## Introduction to the Build a Voting System
```

File: curriculum/challenges/_meta/lab-voting-system/meta.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "Build a Voting System Using Maps and Sets",
```
Comment (Dario-DC) [inline curriculum/challenges/_meta/lab-voting-system/meta.json:2]:
```suggestion
  "name": "Build a Voting System",
```

File: curriculum/challenges/_meta/lab-voting-system/meta.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  "isUpcomingChange": true,
+  "usesMultifileEditor": true,
+  "dashedName": "lab-voting-system",
+  "superBlock": "full-stack-developer",
+  "challengeOrder": [{ "id": "673b567e3ba535dda140d278", "title": "Build a Voting System Using Maps and Sets" }],
```
Comment (Dario-DC) [inline curriculum/challenges/_meta/lab-voting-system/meta.json:7]:
```suggestion
  "challengeOrder": [{ "id": "673b567e3ba535dda140d278", "title": "Build a Voting System" }],
```

File: curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md [hunk 8]
```diff
@@ -0,0 +1,352 @@
+---
+id: 673b567e3ba535dda140d278
+title: Build a Voting System Using Maps and Sets
```
Comment (Dario-DC) [inline curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md:3]:
```suggestion
title: Build a Voting System
```

**Review by Dario-DC — COMMENTED** [review]
In my previous review I forgot to mention that the title should be shortened. I think `Build a Voting System` is enough.

Other than that everything works well :tada:

**Review by Dario-DC — APPROVED** [review]

**Review by moT01 — APPROVED** [review]
LGTM 🎉

**Comment by mitaaios** [comment]
I know u don’t care but it was a person from ca I’m just quit man Sent from my iPhoneOn Dec 9, 2024, at 7:00 PM, Tom ***@***.***> wrote:﻿
Merged #57208 into main.

**Comment by mitaaios** [comment]
213 to now I was right Sent from my iPhoneOn Dec 9, 2024, at 7:00 PM, Tom ***@***.***> wrote:﻿
Merged #57208 into main.

**Comment by mitaaios** [comment]
Was it anyone u know Sent from my iPhoneOn Dec 9, 2024, at 7:00 PM, Tom ***@***.***> wrote:﻿
Merged #57208 into main.

**Pull request merged**